scipy/sparse/csgraph/_flow.pyx

```diff
@@ -407,3 +431,200 @@ cdef ITYPE_t[:] _edmonds_karp(
+        otherwise ``False``.
+
+    """
+    cdef ITYPE_t n_verts = edge_ptr.shape[0] - 1
+
+    cdef ITYPE_t cur, start, end, dst_vertex, e
+
+    q[0] = source
+    start = 0
+    end = 1
+    levels[source] = 0
+
+    while start != end:
+        cur = q[start]
+        start += 1
+        if cur == sink:
+            return 1
+        for e in range(edge_ptr[cur], edge_ptr[cur + 1]):
+            dst_vertex = heads[e]
+            if capacities[e] > 0 and levels[dst_vertex] == -1:
+                levels[dst_vertex] = levels[cur] + 1
+                q[end] = dst_vertex
+                end += 1
+    return 0
+
+cdef ITYPE_t _augment_paths(
+        const ITYPE_t[:] edge_ptr, # IN
+        const ITYPE_t current, # IN
+        const ITYPE_t sink, # IN
+        const ITYPE_t[:] levels, # IN
+        const ITYPE_t[:] heads, # IN
+        const ITYPE_t[:] rev_edge_ptr, # IN
+        ITYPE_t[:] capacities, # IN/OUT
+        ITYPE_t[:] progress, # IN
+        ITYPE_t[:] flows, # OUT
+        ITYPE_t flow # OUT
+    ) nogil:
+    """Computes blocking flow in layered graph using depth first search.
+
+    Parameters
+    ----------
+    edge_ptr : memoryview of length :math:`|V| + 1`
+        For a given vertex ``v``, the edges whose tail is ``v`` are those between
+        ``edge_ptr[v]`` and ``edge_ptr[v + 1] - 1``.
+    current : int
+        The current node in the path.
+    sink : int
+        The sink vertex.
+    levels: memoryview of length :math:`|E|`
+        For a given vertex ``v``, ``levels[v]`` is the level of ``v`` in 
+        the layered graph of input graph.
+    heads : memoryview of length :math:`|E|`
+        For a given edge ``e``, ``heads[e]`` is the head vertex of ``e``.
+    rev_edge_ptr : memoryview of length :math:`|E|`
+        For a given edge ``e``, ``rev_edge_ptr[e]`` is the edge obtained by
+        reversing ``e``. In particular, ``rev_edge_ptr[rev_edge_ptr[e]] == e``.
+    capacities : memoryview of length :math:`|E|`
+        For a given edge ``e``, ``capacities[e]`` is the capacity of ``e``.
+    progress: memoryview of length :math:`|E|`
+        For a given vertex ``v``, ``progress[v]`` is the index of the next
+        edge to be visited from ``v``.
+    flows : memoryview of length :math:`|E|`
+        The residual graph with respect to a maximum flow.
+    flow : int
+        Current maximum flow in the path.
+
+    Returns
+    -------
+    result_flow : int
+        The blocking flow in the layered graph.
+
+    """
+    if current == sink:
+        return flow
+    
+    cdef ITYPE_t dst_vertex, result_flow, e
+
+    if progress[current] == -1:
+        progress[current] = edge_ptr[current]
+    while progress[current] < edge_ptr[current + 1]:
+        e = progress[current]
+        dst_vertex = heads[e]
+        if (capacities[e] > 0 and 
+            levels[dst_vertex] == levels[current] + 1):
+            result_flow = _augment_paths(edge_ptr, dst_vertex, sink,
```

Comment:
Do you have any feeling for how this behaves compared to a stack-based implementation? _Feels_ like it'd be quite a lot of parameter passing for a potentially deep call stack, but maybe it's completely negligible?

Reply:
With the stack implementation in the diff below,
```diff
diff --git a/scipy/sparse/csgraph/_flow.pyx b/scipy/sparse/csgraph/_flow.pyx
index 7dbf94595..6cbd9a928 100644
--- a/scipy/sparse/csgraph/_flow.pyx
+++ b/scipy/sparse/csgraph/_flow.pyx
@@ -495,7 +495,7 @@ cdef bint _build_level_graph(
 
 cdef ITYPE_t _augment_paths(
         const ITYPE_t[:] edge_ptr,  # IN
-        const ITYPE_t current,  # IN
+        const ITYPE_t source,  # IN
         const ITYPE_t sink,  # IN
         const ITYPE_t[:] levels,  # IN
         const ITYPE_t[:] heads,  # IN
@@ -503,7 +503,8 @@ cdef ITYPE_t _augment_paths(
         ITYPE_t[:] capacities,  # IN/OUT
         ITYPE_t[:] progress,  # IN
         ITYPE_t[:] flows,  # OUT
-        ITYPE_t flow  # OUT
+        ITYPE_t flow,  # OUT
+        ITYPE_t[:, :] stack, # IN/OUT
         ) nogil:
     """Computes blocking flow in layered graph using depth first search.
 
@@ -512,8 +513,8 @@ cdef ITYPE_t _augment_paths(
     edge_ptr : memoryview of length :math:`|V| + 1`
         For a given vertex ``v``, the edges whose tail is ``v`` are
         those between ``edge_ptr[v]`` and ``edge_ptr[v + 1] - 1``.
-    current : int
-        The current node in the path.
+    source : int
+        The source vertex.
     sink : int
         The sink vertex.
     levels: memoryview of length :math:`|E|`
@@ -540,29 +541,46 @@ cdef ITYPE_t _augment_paths(
         An arbitrary flow in the layered graph.
 
     """
-    if current == sink:
-        return flow
-
-    cdef ITYPE_t dst_vertex, result_flow, e
-
-    if progress[current] == -1:
-        progress[current] = edge_ptr[current]
-    while progress[current] < edge_ptr[current + 1]:
+    cdef ITYPE_t top, dst_vertex, result_flow, e, current
+    top = 0
+    stack[top][0] = source
+    stack[top][1] = flow
+
+    while top > -1:
+        current = stack[top][0]
+        flow = stack[top][1]
         e = progress[current]
         dst_vertex = heads[e]
         if (capacities[e] > 0 and
                 levels[dst_vertex] == levels[current] + 1):
-            result_flow = _augment_paths(edge_ptr, dst_vertex, sink,
-                                         levels, heads, rev_edge_ptr,
-                                         capacities, progress, flows,
-                                         min(flow, capacities[e]))
-            if result_flow:
-                capacities[e] -= result_flow
-                capacities[rev_edge_ptr[e]] += result_flow
-                flows[e] += result_flow
-                flows[rev_edge_ptr[e]] -= result_flow
-                return result_flow
-        progress[current] += 1
+            if dst_vertex == sink:
+                result_flow = min(flow, capacities[e])
+                if result_flow:
+                    while top > -1:
+                        e = progress[stack[top][0]]
+                        capacities[e] -= result_flow
+                        capacities[rev_edge_ptr[e]] += result_flow
+                        flows[e] += result_flow
+                        flows[rev_edge_ptr[e]] -= result_flow
+                        top -= 1
+                    return result_flow
+                while top > -1:
+                    progress[stack[top][0]] += 1
+                    if progress[stack[top][0]] >= edge_ptr[stack[top][0] + 1]:
+                        top -= 1
+                    else:
+                        break
+            else:
+                top += 1
+                stack[top][0] = dst_vertex
+                stack[top][1] = min(flow, capacities[e])
+        else:
+            while top > -1:
+                progress[stack[top][0]] += 1
+                if progress[stack[top][0]] >= edge_ptr[stack[top][0] + 1]:
+                    top -= 1
+                else:
+                    break
     return 0
 
 cdef ITYPE_t[:] _dinic(
@@ -607,6 +625,7 @@ cdef ITYPE_t[:] _dinic(
     cdef ITYPE_t[:] levels = np.empty(n_verts, dtype=ITYPE)
     cdef ITYPE_t[:] progress = np.empty(n_verts, dtype=ITYPE)
     cdef ITYPE_t[:] q = np.empty(n_verts, dtype=ITYPE)
+    cdef ITYPE_t[:, :] stack = np.empty((n_verts, 2), dtype=ITYPE)
     cdef ITYPE_t[:] flows = np.zeros(n_edges, dtype=ITYPE)
     cdef ITYPE_t flow
 
@@ -617,11 +636,11 @@ cdef ITYPE_t[:] _dinic(
                                   capacities, heads, levels, q):
             break
         for i in range(n_verts):
-            progress[i] = -1
+            progress[i] = edge_ptr[i]
         flow = 1
         while flow:
             flow = _augment_paths(edge_ptr, source, sink,
                                   levels, heads, rev_edge_ptr,
                                   capacities, progress, flows,
-                                  ITYPE_MAX)
+                                  ITYPE_MAX, stack)
     return flows
```
The benchmark results (using the same code as in OP of this PR),

|   Density | Algorithm    |  μ  ±  σ   (SciPy)   |  μ  ±  σ  (MaxFlow)   |
|-----------|--------------|----------------------|-----------------------|
|       0.1 | Dinic        | 0.05 ± 0.0           | 0.016 ± 0.0           |
|       0.1 | Edmonds Karp | 0.052 ± 0.0          | 0.066 ± 0.0           |
|       0.3 | Dinic        | 0.134 ± 0.0          | 0.058 ± 0.0           |
|       0.3 | Edmonds Karp | 0.203 ± 0.0          | 0.719 ± 0.0           |
|       0.5 | Dinic        | 0.22 ± 0.0           | 0.128 ± 0.0           |
|       0.5 | Edmonds Karp | 0.527 ± 0.0          | 1.723 ± 0.0           |
|       0.9 | Dinic        | 0.241 ± 0.0          | 0.097 ± 0.0           |
|       0.9 | Edmonds Karp | 1.225 ± 0.0          | 3.594 ± 0.0           |

I don't see much gains. Anyways, all the parameters in the recursive calls are references (or integers) which amounts to say 88 bytes which is not much as compared to the size of the problem at hand. So, parameter passing overhead is small as far as I can predict. Going for an iterative implementation would be less intuitive IMHO. When, the overall parameter passing approach will be improved then the algorithm would be really fast and stack/recursive implementation won't make much difference.

Reply:
Excellent, thanks for taking the time to make the comparison! I'll update my priors.